pkg/ingester/ingester.go

```diff
@@ -205,14 +209,22 @@ func New(cfg Config, clientConfig client.Config, limits *validation.Overrides, c
 	}
 
 	var err error
-	i.lifecycler, err = ring.NewLifecycler(cfg.LifecyclerConfig, i, "ingester")
+	i.limiter = NewSeriesLimiter(limits, i.lifecycler, cfg.LifecyclerConfig.RingConfig.ReplicationFactor, cfg.ShardByAllLabels)
+
+	i.wal, err = newWAL(cfg.WALConfig, i)
+	if err != nil {
+		return nil, err
+	}
+
+	i.lifecycler, err = ring.NewLifecycler(cfg.LifecyclerConfig, i, "ingester", !cfg.WALConfig.walEnabled)
 	if err != nil {
 		return nil, err
 	}
 
 	// Init the limter and instantiate the user states which depend on it
-	i.limiter = NewSeriesLimiter(limits, i.lifecycler, cfg.LifecyclerConfig.RingConfig.ReplicationFactor, cfg.ShardByAllLabels)
-	i.userStates = newUserStates(i.limiter, cfg)
+	if i.userStates == nil {
```

Comment:
What's the case `i.userStates` is not `nil` at this point?

Reply:
It gets set while recovering the WAL. So if we recover from WAL it's not nil.

Reply:
I would suggest you to add a comment to explain it, so it's easier to understand for people not familiar with the WAL implementation.